agent/proxycfg/api_gateway.go

```diff
@@ -0,0 +1,492 @@
+		})
+
+		snap.APIGateway.TCPRoutes.ForEachKey(func(ref structs.ResourceReference) bool {
+			if _, ok := seenRefs[ref]; !ok {
+				snap.APIGateway.TCPRoutes.CancelWatch(ref)
+			}
+			return true
+		})
+
+		snap.APIGateway.Certificates.ForEachKey(func(ref structs.ResourceReference) bool {
+			if _, ok := seenRefs[ref]; !ok {
+				snap.APIGateway.Certificates.CancelWatch(ref)
+			}
+			return true
+		})
+
+		snap.APIGateway.BoundGatewayConfigLoaded = true
+		break
+	case *structs.APIGatewayConfigEntry:
+		snap.APIGateway.GatewayConfig = gwConf
+
+		for _, listener := range gwConf.Listeners {
+			snap.APIGateway.Listeners[listener.Name] = listener
+		}
+
+		snap.APIGateway.GatewayConfigLoaded = true
+		break
+	default:
+		return fmt.Errorf("invalid type for config entry: %T", resp.Entry)
+	}
+
+	return h.watchIngressLeafCert(ctx, snap)
+}
+
+// handleInlineCertConfigUpdate
+func (h *handlerAPIGateway) handleInlineCertConfigUpdate(ctx context.Context, u UpdateEvent, snap *ConfigSnapshot) error {
+	resp, ok := u.Result.(*structs.ConfigEntryResponse)
+	if !ok {
+		return fmt.Errorf("invalid type for response: %T", u.Result)
+	} else if resp.Entry == nil {
+		return nil
+	}
+
+	cfg, ok := resp.Entry.(*structs.InlineCertificateConfigEntry)
+	if !ok {
+		return fmt.Errorf("invalid type for config entry: %T", resp.Entry)
+	}
+
+	// TODO Consider if unset SectionName and acl.EnterpriseMeta could trip us up
+	ref := structs.ResourceReference{
+		Kind: cfg.GetKind(),
+		Name: cfg.GetName(),
+	}
+
+	snap.APIGateway.Certificates.Set(ref, cfg)
+
+	// TODO
+	return nil
+}
+
+// handleRouteConfigUpdate
+func (h *handlerAPIGateway) handleRouteConfigUpdate(ctx context.Context, u UpdateEvent, snap *ConfigSnapshot) error {
+	resp, ok := u.Result.(*structs.ConfigEntryResponse)
+	if !ok {
+		return fmt.Errorf("invalid type for response: %T", u.Result)
+	} else if resp.Entry == nil {
+		return nil
+	}
+
+	// TODO Consider if unset SectionName and acl.EnterpriseMeta could trip us up
+	ref := structs.ResourceReference{
+		Kind: resp.Entry.GetKind(),
+		Name: resp.Entry.GetName(),
+	}
+
+	seenUpstreamIDs := make(map[UpstreamID]struct{})
+	upstreams := make(map[APIGatewayListenerKey]structs.Upstreams)
+
+	switch route := resp.Entry.(type) {
+	case *structs.HTTPRouteConfigEntry:
+		snap.APIGateway.HTTPRoutes.Set(ref, route)
+
+		for _, rule := range route.Rules {
+			for _, service := range rule.Services {
+				for _, listener := range snap.APIGateway.Listeners {
+					shouldBind := false
+					for _, parent := range route.Parents {
+						if h.referenceIsForListener(parent, listener, snap) {
+							shouldBind = true
+							break
+						}
+					}
+					if !shouldBind {
+						continue
+					}
+
+					upstream := structs.Upstream{
+						DestinationName:      service.Name,
+						DestinationNamespace: service.NamespaceOrDefault(),
+						DestinationPartition: service.PartitionOrDefault(),
+						LocalBindPort:        listener.Port,
+						//IngressHosts:         g.Hosts,
```

Comment:
This is used in the xDS generation code, no?

Reply:
Yes it is, leaving as a TODO (will add TODO since I apparently forgot)